src/QueryManager.ts

```diff
@@ -0,0 +1,222 @@
+import Store from '@orbit/store'
+
+import {
+  Term,
+  OngoingQueries,
+  RecordObject,
+  Expressions,
+  Queries,
+  Subscriptions,
+  QueryRefs,
+  Statuses,
+  QueryCacheOptions,
+} from './types'
+import { Transform, RecordOperation, RecordIdentity } from '@orbit/data'
+import { identityIsEqual } from './helpers'
+
+export class QueryManager<E extends { [key: string]: any } = any>  {
+  _extensions: E
+  _store: Store
+  _ongoingQueries: OngoingQueries
+
+  subscriptions: Subscriptions
+  statuses: Statuses
+
+  constructor (orbitStore: Store, extensions?: E) {
+    this._extensions = extensions || {} as E
+    this._store = orbitStore
+    this._ongoingQueries = {}
+
+    this.subscriptions = {}
+    this.statuses = {}
+  }
+
+  generateQueryRef (queries: Queries) {
+    const terms = this._extractTerms(queries)
+
+    let queryRef = JSON.stringify(terms)
+
+    if (this.subscriptions[queryRef]) return queryRef
+
+    this.subscriptions[queryRef] = { listeners: 0, terms }
+    return queryRef
+  }
+
+  _extractTerms (queries: Queries): Term[] {
+    return Object.keys(queries).sort().map(
+      (key) => ({ key, expression: queries[key](this._store.queryBuilder).expression as Expressions })
+    )
+  }
+
+  query (queryRef: string) {
+    return this._ongoingQueries[queryRef] ? this._ongoingQueries[queryRef].statusRef : this._query(queryRef)
+  }
+
+  _query (queryRef: string) {
+
+    const statusRef = this._generateStatusRef(queryRef)
+    this.statuses[statusRef] = { error: null, loading: false, listeners: 0 }
+
+    const terms = this.subscriptions[queryRef].terms
+
+    const queries: Promise<RecordObject>[] = terms
+      .map(({ key, expression }) =>
+        new Promise((resolve, reject) => {
+          this._store.query(expression)
+            .then(record => resolve({ [key]: record }))
+            .catch(reject)
+        })
+      )
+
+    // The statuses[statusRef] object can be deleted before this gets called if client unsubscribes
```

Comment:
Would be nice if the staus is not deleted until everything is resolved, regardless of amount of listeners.